src/terminal.ts

```diff
@@ -47,6 +58,56 @@ export abstract class Terminal implements ITerminal {
     this._checkType('uid', opt.uid ? opt.uid : null, 'number');
     this._checkType('gid', opt.gid ? opt.gid : null, 'number');
     this._checkType('encoding', opt.encoding ? opt.encoding : null, 'string');
+
+    // setup flow control handling
+    if (opt.handleFlowControl) {
+      this._flowPause = opt.flowPause || FLOW_PAUSE;
+      this._flowResume = opt.flowResume || FLOW_RESUME;
+      this.enableFlowHandling();
+    }
+  }
+
+  /**
+   * Enable automatic flow control handling.
+   * Instead of relying on XON/XOFF directly we pause/resume
+   * the underlying socket. Main advantage is that flow control will
+   * still work even when XON/XOFF is not available. As a downside it
+   * does not act immediately, instead the OS pty buffer and node's socket buffer
+   * fill up first before the slave program gets paused.
+   * Note that `flowPause` and `flowResume` must be in line with the
+   * messages sent by the terminal emulator to indicate PAUSE/RESUME.
+   * @param flowPause String to pause the pty. Defaults to XOFF (x13).
+   * @param flowResume String to resume the pty. Defaults to XON (x11).
+   */
+  public enableFlowHandling(flowPause: string = this._flowPause, flowResume: string = this._flowResume): void {
+    if (this._realWrite) {
+      return;
+    }
+    this._realWrite = this.write;
+    this.write = (data: string) => {
+      // PAUSE/RESUME messages are not forwarded to the pty
+      if (data === flowPause) {
+        this.pause();
+        return;
+      }
+      if (data === flowResume) {
+        this.resume();
+        return;
+      }
+      // normal pty.write
+      this._realWrite(data);
+    };
+  }
+
+  /**
+   * Disable automatic flow control handling.
+   */
+  public disableFlowHandling(): void {
```

Comment:
Toggling this after creation seems like something no one will need?

Reply:
Ah this will be needed for binary transport (for things like zmodem protocol). Binary data tend to contain any sequence by accident, thus ppl need a possibility to disable (and later reenable) it.

Reply:
How would it work at all with zmodem then?

Reply:
zmodem would search for XON/XOFF and escape it with ZDLE char (xor'ed?). Not sure about the implications but I think the control codes are used for its message framing, too. Cannot make much sense from the lines here: https://stackoverflow.com/a/9710116